### ISSUE TITLE
Cannot construct optional tuple elements from spreading 0-1 length Array literals

### ISSUE THREAD

**Issue by pushkine** [microsoft/TypeScript#43257]
[Playground Link](https://www.typescriptlang.org/play?ts=4.3.0-dev.20210315&ssl=4&ssc=32&pln=4&pc=1#code/CYUwxgNghgTiAEEQBd4CNYC54G0DOyMAlgHYDmANPCQK4C2aIMA-ALoDcAsAFCiSwIkqAGYB7UdjTikUEl24YY8ALy4ARGqoA6HTgCMrDj0Ur1m+Dq0AKMaPjNcB7DlYBKDkA)

TS 4.0.5, 4.2.2 and 4.3.0-dev20210314

```ts
declare let bar: [string, number?];
declare let foo: boolean;
bar = ["", ...[]];
bar = ["", ...[1]];
bar = ["", ...(foo ? [1]: [])];
```
```
Type '[string, ...never[]]' is not assignable to type '[string, (number | undefined)?]'.
  Target allows only 2 element(s) but source may have more.(2322)

Type '[string, ...number[]]' is not assignable to type '[string, (number | undefined)?]'.
  Target allows only 2 element(s) but source may have more.(2322)

Type '[string, ...number[]]' is not assignable to type '[string, (number | undefined)?]'.(2322)
```
Unexpected Errors
Expected `...[]` to have no effect when recipient is tuple
Expected `...[1]` to resolve into `number` when recipient is tuple
Expected `(foo ? [1]: [])` to resolve into a union of the above when recipient is tuple

**Comment by captain-yossarian** [comment]
Just for information, it works as expected with `as const`
```typescript
declare let bar: [string, number?];
declare let foo: boolean;

bar = ["", ...[] as const]; // ok
bar = ["", ...[1] as const]; // ok
bar = ["", ...(foo ? [1] as const: [] as const)]; //ok
```
I think this is because TS is unable to figure out length of mutable array `number[]`

**Comment by RyanCavanaugh** [comment]
It'd be nice if this were a const context automatically, but for practical purposes it doesn't really matter since this is only provably of the correct length in places where you wouldn't need to write it as a spread anyway
